---
title: "MOFA: Assessment of the continuity of IGHV factor in the mRNA data"
output:
  BiocStyle::html_document:
    toc: true
---

```{r}
library(MOFAtools)
library(data.table)
library(purrr)
library(magrittr)
library(dplyr)
library(ggplot2)
library(gridExtra)
```



#Load base model
```{r}
load("../out_import_technical.RData")
idx_ighv <- "1"
```

# Load models without drug data 
```{r}
#output path for plots
plotdir <- "~/Documents/MOFA/CLL_MOFA_data/Analysis/code4manuscript/cont_IGHV/Figures"
if(!dir.exists(plotdir)) dir.create(plotdir)

in.folder <- "~/Documents/MOFA/CLL_MOFA_data/out/continuity"

files <- list.files(in.folder, pattern="model_nomet.*\\.hdf5$")
models <- lapply(files, function(fnm) loadModel(file.path(in.folder,fnm), sortFactors = T))
```

# Load patient  data
```{r load_patient_meta, echo=FALSE}
data(patmeta, package="PACEdata")
patmeta <- patmeta[sampleNames(models[[1]]),] %>%
  tibble::rownames_to_column("sample") %>%
  as.data.table %>% .[,c("sample","IGHV","ConsClust")] %>%
  setnames("IGHV","ighv_status") %>%
  setnames("ConsClust","MethylationCluster")
```

# Get our clusters from base model and define colors
```{r}
set.seed(32180)
ZMC <- kmeans(Z[,idx_ighv], 3, nstart=1000, iter.max = 1000)
ZMC <- ZMC$cluster
ZMC <- ifelse(ZMC==1, "HZ", ifelse(ZMC==2, "LZ", "IZ"))
ZMC <- as.data.table(ZMC, keep.rownames=TRUE) %>% setnames("rn", "sample")
col4MC <- c(LZ="navy", IZ="darkgoldenrod", HZ="darkred", "missing"="gray")
patmeta  %<>% merge(ZMC,by="sample") 
```

# Load mRNA data 
```{r}
methData <- read.table("~/Documents/MOFA/CLL_MOFA_data/views/minView=2/meth.txt", header = T)

# Convert to data.table
methData <- methData %>% as.data.table(keep.rownames=T) %>% 
  setnames("rn","sample") %>%
  melt(id.vars="sample", variable.name="CpG", value.name="Mvalue")
```


# Collect IGHV factor from the models (it is always factor 1)
TO-DO: CHECK THE CORRELATION BETWEEN THE FACTOR AND IGHV STATUS
```{r}
ighv_factor <- list()
for (i in 1:length(models)) {
  tmp <- getFactors(models[[i]], factors="1", as.data.frame=T)
  tmp$trial <- i
  tmp %<>% merge(patmeta,by="sample")
  # Align IGHV factors
  fit <- glm(factor(tmp$ighv_status, levels=c("U", "M")) ~ tmp$value, family="binomial")
  if(fit$coefficients[2]<0) tmp %<>% mutate(value = -value)
  
  set.seed(32180)
  ZMC_sub <- kmeans(tmp$value, 3, nstart=1000, iter.max = 1000)
  ZMC_sub <- ZMC_sub$cluster
  ZMC_sub <- ifelse(ZMC_sub==1, "HZ", ifelse(ZMC_sub==2, "LZ", "IZ"))
  tmp$ZMC_sub <- ZMC_sub
  print(sum(tmp$ZMC_sub != tmp$ZMC))
  
  ighv_factor[[i]] <- tmp
}
ighv_factor <- rbindlist(ighv_factor) %>% .[,factor:=NULL]
```


# Scatterplot between IGHV factor and gene expression markers
Only use one trial, as they are all very similar
```{r}
# interesting_CpG <- names(tail(sort(abs(model@Expectations$SW$Methylation$E[,idx_ighv])), n=20))
interesting_CpG <- c("cg17479716","cg19358877","cg26615224")

plist <- lapply(interesting_CpG, function(x) {
  tmp <- filter(methData,CpG==x)
  foo <- merge(tmp,ighv_factor,by="sample", allow.cartesian = T) %>% setnames("value","ighv_factor")
  
  p <- ggplot(filter(foo, trial==1), aes(x=ighv_factor, y=Mvalue, group=trial, shape=as.factor(trial))) +
    geom_point(aes(color=ZMC), size=1.25) +
    stat_smooth(method="lm") +
    theme_bw() + ylab(x) +guides(color=F, shape=F)+
    scale_color_manual(values = col4MC) +xlab("IGHV-Factor")
  p
})

pdf(file.path(plotdir,"Cont_IGHV_noMeth.pdf"), width = 6, height = 2, useDingbats=F)
grid.arrange(plist[[1]],plist[[2]],plist[[3]], ncol=3)
do.call(grid.arrange, plist)
dev.off()
```
