---
title: "Analysis of Factor 1"
author: "Britta Velten"
date: "10/08/2017"
output:
  BiocStyle::html_document:
    toc: true
---
# Introduction
This script analyses the first Factor, which is linked to IGHV status, as shown in Figure 3. Further, it performs imputation of IGHV status based on this factor. The analysis is based on the model and other objects defined in import_models.Rmd.

```{r, message=F, warning=F}
library(MOFAtools)
library(dplyr)
library(reshape2)
library(pheatmap)
library(survival)
library(gridExtra)
library(ggplot2)
library(GGally)
library(magrittr)
library(Hmisc)
library(cowplot)
library(beeswarm)
```

# Get fitted model and related data
(Prepared in import_models.Rmd)
```{r}
#output path for plots and data
plotdir <- "figures/Analysis_Factor1"
if(!dir.exists(plotdir)) dir.create(plotdir)
outdir <- "out"

# load model selected in import_models.Rmd
load(file.path(outdir,"out_import.RData"))

# idx for IGHV latent factor to be analysed in this script
idx_ighv <- "1"

# helper function for some plots
source("plotting_utils.R")
```


# Clustering of patients into 3 groups based on MOFA factor
Based on the value of factor 1 samples are classified into 3 groups using kmeans.
```{r}
# methylation cluster
MC <- covariates[, "MethylationCluster"]
MC[is.na(MC)] <- "missing"
names(MC) <- rownames(covariates)

# IGHV status
IGHV <- covariates[,"IGHV"]
IGHV[is.na(IGHV)] <- "missing"
names(IGHV) <- rownames(covariates)

# kmeans to determine 3 factor clusters
set.seed(32180)
ZMC <- kmeans(Z[,idx_ighv], 3, nstart=1000, iter.max = 1000)
ZMC <- ZMC$cluster
ZMC <- ifelse(ZMC==1, "HZ", ifelse(ZMC==2, "LZ", "IZ"))
table(MC, ZMC)
```


## Beeswarm plot
```{r, fig.width = 3, fig.height = 5}
col4Clusters <- c(LZ="navy", IZ="darkgoldenrod", HZ="darkred")
MCcolors <- col4Clusters[ZMC]

#make plot
# pdf(file.path(plotdir,"BeeswarmZMC.pdf"), width = 3, height = 5, useDingbats=F)
par(mar=c(2.3, 4.5, 4, 2), xpd=TRUE)
bs <- beeswarm(Z[,idx_ighv], pwcol = MCcolors, pch = 16,
          ylab = paste("Factor", idx_ighv), xlab = "",
          cex.lab=1.5, cex.axis=1.5, cex.main=1.5, cex.sub=1.5)

legend("top", legend = c("LZ", "IZ", "HZ"),
       title = "Factor clusters", pch = 16,
       col = col4Clusters, ncol = 3,
       inset=c(0,-0.2), cex=1.2,
       box.lwd = 0, box.col = "white")
# dev.off()    
```



# Characterization in the mRNA view

## Top Weights
```{r,fig.width=5.6, fig.height=8}
# number of top genes to show
ntop_mRNA <- 20

# list of previously mentioned IGHV associated genes (see references in ms)
knownGenes <- c("AKAP12", "ADAM29", "BCL7A", "CLECSF2", "FCGBP", "FLJ10884",
                "FUT8", "LPL", "TCF7", "WNT3", "APOD", "SPG20", "MYL9", "NRIP1",
                "SPAP1", "SPRY2", "TGFBR3", "ZAP70", "COBLL1", "ZNF667", "SEPT10",
                "CRY1", "PLD1", "BCL7A", "WNT9A")

# plot top weights
# pdf(file.path(plotdir,"mRNAweigths_IGHV.pdf"),width=5.6, height=8, useDingbats = FALSE)
p <- showTopWeightsAndColor(model, "mRNA", idx_ighv,
                            nfeatures = ntop_mRNA,
                            Features2color = knownGenes,
                            scalePerView = T,
                            col2highlight = "darkorange",
                            orderBySign = TRUE)
p
# dev.off()                                                                                                                               
```

## Data Heatmap
```{r, fig.width=7, fig.height=8}
# get top genes and patient having RNAseq
topGenes <- names(sort(abs(model@Expectations$SW$mRNA$E[,idx_ighv]), decreasing = T))[1:ntop_mRNA]
topGenes <- topGenes[order(model@Expectations$SW$mRNA$E[topGenes,idx_ighv], decreasing = T)]
patients2include <- colnames(data$mRNA)[apply(data$mRNA,2, function(p) !any(is.na(p)))]

# annotate by IGHV factor 
anno_df <- data.frame(Z=Z[,idx_ighv], ZMC=ZMC)
colnames(anno_df) <- c("Factor 1", "Clusters")
rownames(anno_df) <- rownames(Z)
annoHM_colors <- list(c("blue", "red"), col4Clusters)
names(annoHM_colors) <- c("Factor 1", "Clusters")

# heatmap
# pdf(file.path(plotdir,"heatmap_mRNA_MC_rotated2.pdf"),width=7, height=8, onefile = F)
pheatmap(data$mRNA[topGenes, patients2include[order(Z[patients2include, idx_ighv])] ],
         show_colnames = F, cluster_rows = F, cluster_cols = F, fontsize = 18,
         annotation_col = anno_df, annotation_legend = F,
         gaps_col = c(which(ZMC[order(Z[patients2include,idx_ighv], decreasing = F)]=="IZ")[1]-1,
                      which(ZMC[order(Z[patients2include,idx_ighv], decreasing = F)]=="HZ")[1]-1),
         show_rownames = T, legend = T, annotation_colors = annoHM_colors)
# dev.off()
```



# Characterization in the drug response view
## Top Weights
```{r, fig.width=7, fig.height=5}
# build df for drug weights
dfDrugs <- as.data.frame(getWeights(model, "Drugs", idx_ighv))
colnames(dfDrugs) <- "loadings"

#scale loadings
dfDrugs$loadings <- dfDrugs$loadings/max(abs(dfDrugs$loadings))

# get concentration, drug and target category
dfDrugs$drug <- substr(rownames(dfDrugs),1,nchar(rownames(dfDrugs))-2)
dfDrugs$conc <- substr(rownames(dfDrugs),nchar(rownames(dfDrugs)),nchar(rownames(dfDrugs)))
dfDrugs$drugid <- rownames(drugs)[match(dfDrugs$drug, drugs[,"name"])]
dfDrugs %<>% mutate(abs_loadings = abs(loadings))
dfDrugs %<>% mutate(target_category = drugs[drugid,"target_category"])
dfDrugs %<>% mutate(pathway = drugs[drugid,"pathway"])
dfDrugs %<>% mutate(main_targets = drugs[drugid,"main_targets"])

# take average across concentrations
dfDrugsAv <- dfDrugs %>% group_by(drug) %>% 
  dplyr::summarise(meanloadings = mean(loadings),
            target_category=unique(target_category),
            main_targets=unique(main_targets),
            pathway=unique(pathway))
dfDrugsAv$absMeanloading <- abs(dfDrugsAv$meanloadings)

# make broader target categories      
dfDrugsAv %<>% mutate(main_targets2 = ifelse(grepl("CHK",main_targets),
                                             "CHK", main_targets))
dfDrugsAv %<>% mutate(main_targets2 = ifelse(grepl("PI3K|SYK|BTK",main_targets2),
                                             "BCR pathway", main_targets2))
dfDrugsAv %<>% mutate(main_targets2 = ifelse(grepl("AKT|LYN|SRC",main_targets2),
                                             "BCR pathway", main_targets2))
dfDrugsAv %<>% mutate(main_targets2 = ifelse(!main_targets2 %in% c("BCR pathway", "CHK", "HSP90"),
                                             "other", main_targets2))

# define colors
cols <- RColorBrewer::brewer.pal(10,"Paired")[c(1,3,9)]
names(cols) <-  c("BCR pathway", "CHK" , "HSP90")
col4Categories <- c(other="gray",cols)

# Sort according to loadings
ntop <- 15
dfDrugsAv <- dfDrugsAv[order(dfDrugsAv$absMeanloading, decreasing = T)[1:ntop],]
dfDrugsAv$drug <- factor(dfDrugsAv$drug, levels=rev(dfDrugsAv$drug))

# Make plot
# pdf(file.path(plotdir,"drugweigths_IGHV.pdf"),width=7, height=5, useDingbats=F)
ggplot(dfDrugsAv,aes(x=drug, y=absMeanloading, col= main_targets2)) +
    geom_point(size=3) +
    geom_segment(aes(xend=drug, yend=0), size=2) +
    coord_flip() +
    theme(
      axis.title.x = element_text(size=rel(1.3), color='black'),
      axis.title.y = element_blank(),
      axis.text.y = element_text(size=rel(1.5), hjust=1, color='black'),
      axis.text.x = element_text(size=rel(1.5), color='black'),
      axis.ticks.y = element_blank(),
      axis.ticks.x = element_line(),
      legend.position='right',
      legend.title=element_text(size=rel(1.3), color="black"),
      legend.text=element_text(size=rel(1.3), color="black"),
      legend.key=element_rect(fill='transparent'),
      panel.background = element_blank()
    ) + ylab("Absolute loading on factor 1") +
  guides(color=guide_legend(title="Categories", nrow=4, title.position="top")) +
  scale_color_manual(values=col4Categories)
# dev.off()
```




## Drug Response Curves
```{r, fig.width = 6, fig.height = 5}
  data(conctab, package="pace")
  groups <- ZMC
  groupnm <- "clusters"
  drugResDF <- lapply(c("dasatinib", "AZD7762"), function(drugnm) {
    drugData2plot <-model@TrainData$Drugs[grepl(drugnm,rownames(model@TrainData$Drug)),]
    drugid <- rownames(drugs[drugs$name==drugnm, ])
  
    drugDF <- melt(drugData2plot, varnames = c("drug", "patient"), value.name = "viability")
    drugDF %<>% mutate(concentrationID = as.numeric(sapply(as.character(drug), function(x) strsplit(x, "_")[[1]][2])))
    drugDF %<>% mutate(concentration = as.numeric(conctab[drugid,paste0("c", concentrationID)]))
    if(!is.null(groups)) drugDF %<>% mutate(group = as.factor(groups[patient])) else drugDF$group <- factor(1)

    drugDF %<>% filter(!is.na(viability) & !is.na(group))
    summary_drugDF <-  drugDF %>% group_by(group, concentrationID, concentration) %>%
            dplyr::summarize(mean_viab = mean(viability), sd = sd(viability), n = length(viability))
    summary_drugDF$se <- summary_drugDF$sd/sqrt(summary_drugDF$n)
    summary_drugDF$drug <- drugnm
    summary_drugDF
}) %>% bind_rows()
  
  p <- ggplot(drugResDF, aes(x=concentration, y=mean_viab, col=group, grou=group)) +
    geom_errorbar(aes(ymin=mean_viab-2*se, ymax=mean_viab + 2*se), width=0.1)+ geom_line(size=2) +
    ylab("viability") +theme_bw(base_size = 21) + facet_wrap(~drug)+
    xlab(expression(paste("Concentration [",mu,"M]"))) #+ scale_x_reverse()
  if(is.null(groups)) p <- p + guides(col=F) else p <- p + guides(col=guide_legend(title =groupnm))
  
  p <- p + scale_color_manual(values = col4Clusters, labels=c("LZ", "IZ", "HZ")) +
    ylim(c(0,1.05)) + theme(legend.position = "top", axis.text = element_text(colour="black")) 
  
  # pdf(file.path(plotdir, paste0("DrugResponseCurve_joint.pdf")), width = 6, height = 5, onefile = F)
  print(p )
  # dev.off()
  
```


# Beeswarm plots for Methylation Cluster
Beesawrm plot of factor 1 with samples colored based on their Methylation Cluster. (To compare with the clusters based on factor 1)
```{r, fig.width = 3, fig.height = 5}
# get methylation cluster and define colors
MC <- covariates[,"MethylationCluster"]
MC[is.na(MC)] <- "missing"
MC <- factor(MC, levels = c("LP", "IP", "HP", "missing"))
col4MC <- c(LP="navy", IP="darkgoldenrod", HP="darkred", "missing"="gray")
MCcolors <- col4MC[MC]

#make plot
# pdf(file.path(plotdir,"BeeswarmMC.pdf"), width = 3, height = 5, useDingbats=F)
par(mar=c(2.3, 4.5, 4, 2), xpd=TRUE)
bs <-beeswarm(Z[,idx_ighv], pwcol = MCcolors, pch = 16,
          ylab = paste("Factor", idx_ighv), xlab = "",
          cex.lab=1.5, cex.axis=1.5,
          cex.main=1.5, cex.sub=1.5)

legend("top", legend =  levels(MC)[1:3],
       title = "Methylation Cluster", pch = 16,
       col = col4MC[1:3], ncol = 4,
       inset=c(0,-0.2), cex=1.2,
       box.lwd = 0, box.col = "white")
# dev.off()    
```

#SessionInfo
```{r}
sessionInfo()
```