---
title: "Load MOFA models and first inspections"
author: "Britta Velten"
date: "10/08/2017"
output:
  BiocStyle::html_document:
    toc: true
---

# Introduction
This script contains the code to load all the trained MOFA models on the CLL data, compare them and pick one model for all further analysis based on the ELBO. The output of this script ('out_import.RData') is used in all other analysis scripts

```{r, message=F, warning=F}
library(MOFAtools)
library(pace)
```

# I/O - Options
```{r}
#output path for plots and data
plotdir <- "figures/import_models"
outdir <- "out"
if(!dir.exists(plotdir)) dir.create(plotdir)
if(!dir.exists(outdir)) dir.create(outdir)

# paths to data
dataDir <- "../data"
# output directory containing all trained MOFA models
mofaDir <- "../data/MOFA_fits"
```

# Load fitted models
In total, 25 models are fitted on patients with at least two views observed (n=200). k (=number of factors) is learnt, intercept term is learned. For details see run_MOFA/run_multiple.sh.
The following views are included: Mutations (>= 3 occurences), Methylation (top 1%, no XY), mRNA (top 5000, no Y), Drugs (all).
```{r, echo=FALSE}
detectMOFAPassengers <- function(object, views = "all", factors = "all", r2_threshold = 0.03) {
  
  # Sanity checks
  if (class(object) != "MOFAmodel") stop("'object' has to be an instance of MOFAmodel")
  
  # Define views
  if (paste0(views,sep="",collapse="") =="all") { 
    views <- viewNames(object) 
  } else {
    stopifnot(all(views %in% viewNames(object)))  
  }
  M <- length(views)
  
  # Define factors
  factors <- as.character(factors)
  if (paste0(factors,collapse="")=="all") { 
    factors <- factorNames(object) 
  } else {
    stopifnot(all(factors %in% factorNames(object)))  
  }
  
  # Collect relevant data
  Z <- getFactors(object)
  
  # Identify factors unique to a single view
  r2 <- calculateVarianceExplained(object, views = views, factors = factors, plotit = F, totalVar = F)$R2PerFactor
  unique_factors <- names(which(rowSums(r2>=r2_threshold)==1))
  
  # Mask samples that have full missing views
  missing <- sapply(getTrainData(object,views), function(view) MOFAtools::sampleNames(object)[apply(view, 2, function(x) all(is.na(x)))] )
  names(missing) <- viewNames(object)
  
  
  for (factor in unique_factors) {
    # view <- names(which(r2[factor,]>=r2_threshold))
    view <- colnames(r2[,which(r2[factor,]>=r2_threshold),drop=F])
    missing_samples <- missing[[view]]
    if (length(missing_samples)>0) {
      Z[missing_samples,factor] <- NA
    }
  }
  object@Expectations$Z$E <- Z
  
  return(object)
  
}
```

```{r}
files <- list.files(mofaDir)
files <- files[grepl(".hdf5",files)]

# load models, detect passengers and sort factors by R2
AllModels <- lapply(files, function (filenm){
  modeltmp <- loadModel(file.path(mofaDir, filenm), sortFactors = T)
  
  #remove passengers
  detectMOFAPassengers(modeltmp)
})

names(AllModels) <- sub(".hdf5","", files)
```

# Preparations
Set nice names
```{r}
# get gene annotations from ENSEMBL
mRNA_file <- file.path(dataDir,"Hsapiens_genes_BioMart.75.txt")
mRNA = read.csv(file=mRNA_file,header=T,sep="\t",stringsAsFactors=F)

# get drug annotations from pace package
data("drugs", package = "pace")

# set feautre names and factor names
AllModels <- lapply(AllModels, function(model){
MOFAtools::featureNames(model)$mRNA <- mRNA$symbol[match(MOFAtools::featureNames(model)$mRNA, mRNA$ens_id)]
MOFAtools::featureNames(model)$Drugs <- paste(drugs[substr(MOFAtools::featureNames(model)$Drugs,1,5),"name"],
                                              substr(MOFAtools::featureNames(model)$Drugs,6,8), sep="")
MOFAtools::featureNames(model)$Mutations[MOFAtools::featureNames(model)$Mutations=="del13q14_any"] <- "del13q14"
MOFAtools::factorNames(model)[grepl("^0",MOFAtools::factorNames(model))] <- "intercept"
model
})
```

In robustness scripts the robustness to initilizations across models is tested and ELBO for the different models is visualized. (Removed those parts here)

# Pick a model based on highest ELBO
```{r}
modelidx <- which.max(sapply(AllModels, function(model) model@TrainStats$elbo[length(model@TrainStats$elbo)]))
modelidx
model <- AllModels[[modelidx]]
```

# Flip factor 1 (for alignment with IGHV in terms of clinical outcome)
Flip sign of latent factors to have a consistent hazard ratio for IGHV and its LF.
```{r}
flip_factor <- function(factor, model){
  model@Expectations$Z$E[,factor] <- - model@Expectations$Z$E[,factor]
  for(viewnm in names(model@Expectations$SW)) {
    model@Expectations$SW[[viewnm]]$E[,factor] <- -model@Expectations$SW[[viewnm]]$E[,factor]
    model@Expectations$SW[[viewnm]]$EW[,factor] <- -model@Expectations$SW[[viewnm]]$EW[,factor]
  }
return(model)
}

model <- flip_factor("1", model)
```

# Extracting important parts of the model and covaraites
Get relevant parts from model
```{r}
data <- model@TrainData
k <- model@Dimensions$K
Z <- getExpectations(model, "Z", "E")
weights <- getExpectations(model, "SW", "E")
```

Get other relevant covariates
```{r}
data("patmeta", package = "pace")
recurMuts <- t(data$Mutations[rowSums(data$Mutations, na.rm = T) >=5, ])
covariates <- cbind(recurMuts[,!grepl("IGHV", colnames(recurMuts))], patmeta[rownames(Z),])
colnames(covariates)[grep("Gender",colnames(covariates) )] <- "sex"
covariates$sex <- ifelse(covariates$sex == "m", 0,1)
colnames(covariates)[grep("Age4Main",colnames(covariates) )] <- "age"
colnames(covariates)[grep("ConsClust",colnames(covariates) )] <- "MethylationCluster"
covariates <- covariates[, !grepl("T5|T6|Age4Pilot|Age4Main|Diagnosis|died|treated|treatedAfter", colnames(covariates) )]
covariates$missingViews <- rowSums(sapply(data, function(dat) apply(is.na(dat), 2, all)))
```

# Correlation of latent factors
```{r, fig.width = 8, fig.height = 7}
# pdf(file.path(plotdir,"factor_correlation.pdf"), onefile = F, width = 8, height = 7)
fout <- FactorsCorPlot(model)
# dev.off()
```

# Save workspace to be used in other down-stream analysis scripts
```{r}
rm(plotdir)
save("model", "covariates", "weights", "data", "k", "Z", file=file.path(outdir,"out_import.RData"))
```

# SessionInfo
```{r}
sessionInfo()
```

