---
title: "MOFA applied to the scMT data"
output:
  BiocStyle::html_document: 
    toc: false
    fig_width: 10
    fig_height: 8
---

```{r load_modules, echo=FALSE, include=FALSE}
devtools::load_all("/Users/ricard/mofa/MOFAtools")
# library(MOFAtools)
library(data.table)
library(purrr)
library(ggplot2)
library(scater)

matrix.please <- function(x) {
    m<-as.matrix(x[,-1])
    rownames(m)<-x[[1]]
    m
}

impute <- function(d, margin) {
  if (margin == 1)
    means <- rowMeans(d, na.rm=T)
  else if (margin == 2)
    means <- colMeans(d, na.rm=T)
  else
    stop("Margin has to be either 1 (rows) or 2 (cols)")
  
  if (any(is.na(means))) {
    stop('Insufficient data for mean imputation!')
  }
  
  for (i in 1:length(means)) {
    if (margin == 1)
      d[i,is.na(d[i,])] <- means[i]
    else if (margin == 2)
      d[is.na(d[,i]), i] <- means[i]
  }
  return (d)
}
```

<!-- Define I/O and options -->
```{r define_opts, echo=FALSE}

## Define I/O ##
io <- list()

io$sample.metadata <- "/Users/ricard/data/scMT/sample_info.txt"
io$met.dir <- "/Users/ricard/data/scMT/met/raw"
io$rna.file <- "/Users/ricard/data/scMT/expr/processed/sceset.rds"

io$annos_dir  <- "/Users/ricard/data/scMT/features/filt"
io$outdir <- "/Users/ricard/mofa_rebuttal/mofa_nmt/out"

## Define options ##
opts <- list()

# Define which annotations to look at
opts$met.annos <- c("CGI","prom_2000_2000","activeEnhancers_ENCODE")

# Define which stage and lineages to look at 
opts$culture <- c("2i","Serum")

# Filtering options for methylation
opts$met_min.coverage <- 0.1 # minimum coverage (pct across cells)
opts$met_nfeatures <- 5000   # maximum number of features per view (filter based on variance)

# Filtering options for RNA
opts$rna_rm.low <- T       # Remove low expressed genes
opts$rna_rm.const <- T     # Remove non-variable genes
opts$rna_min.cdr <- 0.10   # Remove genes with cellular detection rate smaller than opts$min.cdr
opts$rna_ngenes <- 5000    # maximum number of genes (filter based on variance)

# (...)
opts$onlyOverlappingCells <- F

# Define which cells to use
opts$met_cells <- fread(io$sample.metadata) %>% .[pass_metQC==T & culture%in%opts$culture,sample]
opts$rna_cells <- fread(io$sample.metadata) %>% .[pass_rnaQC==T & culture%in%opts$culture,sample]
if (opts$onlyOverlappingCells == T) {
  opts$met_cells <- opts$rna_cells <- Reduce(intersect, list(opts$met_cells,opts$rna_cells))
}
```

<!-- Load sample metadata -->
```{r load_metadata}
sample_metadata <- fread(io$sample.metadata)[,c("sample","culture","pass_rnaQC","pass_metQC")]
```

```{r}
anno_dt <- lapply(opts$met.annos, function(anno) fread(sprintf("%s/%s.bed", io$annos_dir, anno), 
    colClasses = c("character","integer","integer","character","character","character"))[, c(1,2,3,4,5,6)]) %>%
    rbindlist %>% setnames(c("chr","start","end","strand","id","anno")) %>% .[,c("anno","chr"):=list(as.factor(anno), as.factor(paste0("chr",chr)))]
```

<!-- Load methylation data -->
```{r load_data, echo=FALSE, include=FALSE}
met_dt <- lapply(opts$met_cells, function(x) {
  data <- fread(sprintf("zcat < %s/%s.tsv.gz",io$met.dir,x)) %>% setnames(c("chr","pos","rate")) %>%
    .[,sample:=x] %>% .[,chr:=paste0("chr",chr)]
  
  # Filter genomic contexts
  data[,c("start","end") := list(pos,pos)] %>% .[,chr:=as.factor(chr)] %>% setkey(chr,start,end)
  data <- foverlaps(data, anno_dt%>%setkey(chr,start,end), nomatch=0, mult="first") %>%
    .[,c("start","end","strand","i.start","i.end"):=NULL]
  
  return(data)
}) %>% rbindlist
```

<!-- Parse methylation data -->
```{r}
met_dt[,chr_pos:=paste(chr,pos,sep="_")]
met_dt <- merge(met_dt, sample_metadata[,c("sample","culture")], by="sample")
```

<!-- Load RNA data -->
```{r}
sce <- readRDS(io$rna.file)
```

<!-- Parse RNA expression data -->
```{r}
# Filter cells
sce <- sce[,colnames(sce) %in% opts$rna_cells]

# Convert to data.table
rna_dt <- exprs(sce) %>% t %>% as.data.table(keep.rownames = "sample") %>%
  melt(id.vars = "sample", value.name = "expr", variable.name = "gene") %>%
  # merge(rowData(sce) %>% as.data.frame(row.names = rownames(sce)) %>% tibble::rownames_to_column("gene") %>% .[,c("gene","ens_id")] %>% setnames("ens_id","id"))
  merge(fData(sce) %>% as.data.frame(row.names = rownames(sce)) %>% tibble::rownames_to_column("gene") %>% .[,c("gene","ens_id")] %>% setnames("ens_id","id"))

rna_dt <- merge(rna_dt, sample_metadata[,c("sample","culture")], by="sample")
```

<!-- Filter RNA expression data -->
```{r}

# Remove lowly expressed genes
if (opts$rna_rm.low) {
  rna_dt <- rna_dt[,mean:=mean(expr),by="id"] %>% .[mean>=1.0] %>% .[,mean:=NULL]
}

# Remove genes with constant expression levels and sites with constant methylation levels
if (opts$rna_rm.const) {
  rna_dt <- rna_dt[,var:=var(expr),by="id"] %>% .[var>0] %>% .[,var:=NULL]
}

# Filter genes with low cellular detection rate and sites with low coverage across samples
rna_dt <- rna_dt[,cdr:=sum(expr>0)/length(opts$rna_cells), by="gene"] %>% .[cdr>=opts$rna_min.cdr] %>% .[,cdr:=NULL]

# Extract top N highly variable genes
keep_hv_genes <- rna_dt[,.(var=var(expr)), by="gene"] %>% setorder(-var)  %>% head(n = opts$rna_ngenes) %>% .$gene
rna_dt <- rna_dt[gene%in%keep_hv_genes]
```

<!-- Filter methylation data -->
```{r filter_data}

# Filter features by coverage
for (s in unique(met_dt$culture)) {
  met_dt[culture==s,Ntotal:=sample_metadata[sample%in%opts$met_cells & culture==s,.N]]
}
keep_cov_sites <- met_dt %>% split(.$culture) %>% map(~ .[, cov:=.N/Ntotal, by=c("chr_pos","anno")] %>% .[cov >= opts$met_min.coverage] %>% .$chr_pos)
met_dt <- met_dt %>% .[chr_pos%in%Reduce("intersect",keep_cov_sites)] %>% .[,"Ntotal":=NULL]

# Filter features by variance
keep_hv_sites <- met_dt %>% split(.$anno) %>% map(~ .[,.(var = var(rate)), by="chr_pos"] %>% .[var>0] %>% setorder(-var) %>% head(n = opts$met_nfeatures) %>% .$chr_pos)
met_dt <- met_dt %>% split(.$anno) %>% map2(.,names(.), function(x,y) x[chr_pos %in% keep_hv_sites[[y]]]) %>% rbindlist
```


<!-- Create matrix from the data.table -->
```{r}
met_cells <- unique(met_dt$sample)
rna_cells <- unique(rna_dt$sample)

rna_matrix <- rna_dt[,c("gene","expr","sample")] %>%
  .[,sample:=factor(sample,levels=Reduce(union,list(rna_cells,met_cells)))] %>%
  dcast(sample~gene, value.var="expr", drop=F) %>% matrix.please() %>% t

met_matrix_list <- list()
for (n in unique(met_dt$anno)) {
  met_matrix_list[[paste("met",n,sep="_")]] <- met_dt[anno==n,c("chr_pos","rate","sample")] %>%
    .[,sample:=factor(sample,levels=Reduce(union,list(rna_cells,met_cells)))] %>%
    dcast(sample~chr_pos, value.var="rate", drop=F) %>% matrix.please() %>% t
}

all_matrix_list <- c(rna=list(rna_matrix),met_matrix_list)
```

<!-- Create MultiAssayExperiment object -->
```{r}

# Create ExpressionSet for RNA data
sample_metadata_filt <- sample_metadata[pass_rnaQC==T,c("sample","culture")] %>% as.data.frame
rna_matrix_filt <- rna_matrix[,sample_metadata_filt$sample] %>% round(3)
stopifnot(colnames(rna_matrix_filt)==sample_metadata_filt$sample)
rna.pdata <- sample_metadata_filt %>% tibble::column_to_rownames("sample") %>% as("AnnotatedDataFrame")
rna.pdata$cellular_detection_rate <- apply(rna_matrix_filt,2,function(x) sum(x>0))
rna.fdata <- data.frame(row.names = rownames(rna_matrix_filt)) %>% as("AnnotatedDataFrame")
rna.expressionset <- ExpressionSet(assayData=rna_matrix_filt, phenoData=rna.pdata, featureData=rna.fdata)

# Create ExpressionSet for Methylation data
met.expressionset <- list()
sample_metadata_filt <- sample_metadata[pass_metQC==T,c("sample","culture")] %>% as.data.frame
for (i in names(met_matrix_list)) {
  met_matrix_filt <- met_matrix_list[[i]][,sample_metadata_filt$sample] %>% round(3)
  stopifnot(colnames(met_matrix_filt)==sample_metadata_filt$sample)
  met.pdata <- sample_metadata_filt %>% tibble::column_to_rownames("sample") %>% as("AnnotatedDataFrame")
  met.pdata$mean_met <- apply(met_matrix_filt,2,mean, na.rm=T)
  met.fdata <- data.frame(row.names = rownames(met_matrix_filt)) %>% as("AnnotatedDataFrame")
  met.fdata$mean <- apply(met_matrix_filt,1,mean, na.rm=T)
  met.fdata$coverage <- apply(met_matrix_filt,1,function(x) mean(!is.na(x)))
  met.fdata$chr <- sapply(strsplit(rownames(met.fdata),"_"),"[[",1)
  met.fdata$position <- sapply(strsplit(rownames(met.fdata),"_"),"[[",2)
  met.expressionset[[i]] <- ExpressionSet(assayData=met_matrix_filt, phenoData=met.pdata, featureData=met.fdata)
}

# Create SampleMap

sample_metadata_filt <- sample_metadata[pass_rnaQC==T,c("sample","culture")] %>% as.data.frame
rna.map <- data.frame(primary=sample_metadata_filt$sample, colname=sample_metadata_filt$sample, stringsAsFactors = FALSE)

sample_metadata_filt <- sample_metadata[pass_metQC==T,c("sample","culture")] %>% as.data.frame
met.map <- data.frame(primary=sample_metadata_filt$sample, colname=sample_metadata_filt$sample, stringsAsFactors = FALSE)

listmap <- list(rna.map, met.map, met.map, met.map)
names(listmap) <- names(all_assays) <- c("RNA expression","Met Enhancers","Met CpG Islands","Met Promoters")
sampleMap <- MultiAssayExperiment::listToMap(listmap)


# Create MultiAssayExperiment
all_assays <- c(list(rna.expressionset),met.expressionset)
names(all_assays) <- c("RNA expression","Met Enhancers","Met CpG Islands","Met Promoters")
sample_metadata_filt <- sample_metadata %>% as.data.frame %>% tibble::column_to_rownames("sample")
MAE <- MultiAssayExperiment::MultiAssayExperiment(experiments=all_assays, 
                            colData = as.data.frame(sample_metadata) %>% tibble::column_to_rownames("sample"),
                            sampleMap = sampleMap
                            )
scMT_data <- MAE
save(data=scMT_data, file="/Users/ricard/mofa/MOFAtools/data/scMT_data.RData")
load("/Users/ricard/mofa/MOFAtools/data/scMT_data.RData")
```

<!-- Run MOFA -->
```{r}
outfile <- paste0(io$outdir,"/model_test.hdf5")

# MOFAobject <- createMOFAobject(all_matrix_list)
MOFAobject <- createMOFAobject(list(all_matrix_list$rna))
DirOptions <- list("dataDir" = tempdir(), "outFile" = outfile)
DataOptions <- getDefaultDataOpts()
ModelOptions <- getDefaultModelOpts(MOFAobject)
ModelOptions$numFactors <- 10
ModelOptions$sparsity <- T
ModelOptions$learnIntercept <- T
# ModelOptions$covariates <- data.frame(mean=sample_metadata$mean)
TrainOptions <- getDefaultTrainOpts()
TrainOptions$maxiter <- 2000
TrainOptions$tolerance <- 0.00
TrainOptions$learnFactors <- F
TrainOptions$DropFactorThreshold <- 0.00
MOFAmodel <- prepareMOFA(MOFAobject, DirOptions = DirOptions, DataOptions = DataOptions, ModelOptions = ModelOptions, TrainOptions = TrainOptions)
model <- runMOFA(MOFAmodel, DirOptions, mofaPath = "/Users/ricard/anaconda2/bin/mofa")
```

```{r}
outfile <- paste0(io$outdir,"/model.hdf5")
model <- loadModel(outfile)
```

<!-- Parse the MOFA model -->
```{r}

# Filter sample metadata
sample_metadata_filt <- sample_metadata %>% setkey(sample) %>% .[MOFAtools::sampleNames(model)]

# Rename views
viewNames(model) <- stringr::str_replace_all(viewNames(model), 
                                             c("met_CGI"="Met CpG Islands",
                                               "met_activeEnhancers_ENCODE"="Met ESC enhancers",
                                               "met_prom_2000_2000"="Met Promoters",
                                               "rna" = "RNA expression"))
```

<!-- Plot data summary -->
```{r}
pdf(file=paste0(io$outdir,"/TilesData.pdf"), height=5.5, width=7)
plotTilesData(model, colors=c("#377EB8","#377EB8","#377EB8","#31A354"))
dev.off()
```

Scree plot
```{r}
r2 <- calculateVarianceExplained(model)$R2PerFactor
r2[r2<0] <- 0


foo <- r2 %>% melt %>% setnames(c("factor","view","r2"))
# foo <- apply(r2,1,sum) %>% melt %>% as.data.table %>% .[,factor:=as.character(1:nrow(.))]
p <- ggplot(foo, aes(x=factor, y=r2)) +
  geom_point() +
  facet_wrap(~view, ncol=1, scales="free_y") +
  geom_vline(xintercept = 3.5, linetype = "dashed") +
  # geom_hline(yintercept = 0.01, linetype = "dashed") +
  theme_bw()
print(p)
```

```{r}
plotWeights(model, view="rna", factor=4)
plotFactorCor(model)
```

<!-- Plot variance explained -->
```{r}
r2 <- calculateVarianceExplained(model)

pdf(file=paste0(io$outdir,"/varianceExplained.pdf"))
plotVarianceExplained(model)
dev.off()
```

<!-- Plot correlation between factors -->
```{r}
# plotFactorCor(model)
```

<!-- Plot Factor values -->
```{r}
pdf(file=paste0(io$outdir,"/scatterPlot.pdf"), useDingbats = F, width=6, height=4)
plotFactorScatter(model, factors=c("1","2"), color_by=sample_metadata_filt$culture) +
  scale_color_manual(values=c("lightsalmon","orangered3"))
  # scale_shape_manual(values=c(19,1,2:18)[1:length(unique(shape_by))]) +
dev.off()

# pdf(file=paste0(io$outdir,"/beeswarmPlot.pdf"), useDingbats = F)
# plotFactorBeeswarm(model, factors=c("1"), color_by=sample_metadata_filt$culture)
# dev.off()
```

<!-- Characterise Latent Factor 1 -->
```{r}
# Plot top weights of RNA view
pdf(file=paste0(io$outdir,"/topWeights_Factor1.pdf"), useDingbats = F, width=4.5, height=1.5)
plotWeights(model, view="RNA expression", factor=1, nfeatures = 0, abs=T, 
            manual=list(c("Zfp42","Esrrb","Morc1","Fbxo15","Jam2","Klf4","Tcl1","Tbx3","Tex19.1")))
dev.off()

# Plot RNA top features heatmap
df <- data.frame(
  row.names = sample_metadata_filt$sample,
  culture = sample_metadata_filt$culture
)
pdf(file=paste0(io$outdir,"/Heatmap_Factor1.pdf"), width=6, height=4, onefile = F)
# df <- data.frame(
#   row.names = sample_metadata_filt$sample,
#   culture = sample_metadata_filt$culture
# )
# plotDataHeatmap(model, view="RNA expression", factor=1, features = 100, annotation_row=df, show_colnames=F, show_rownames=F)

setkey(sample_metadata_filt,sample)
z <- sort(getFactors(model,"1")[,1])
df <- data.frame(
  row.names = names(z),
  culture = sample_metadata_filt[names(z)]$culture,
  factor = z
)
plotDataHeatmap(model, view="RNA expression", factor=1, features = 20, show_colnames=F, show_rownames=T, transpose = T, sortSamples = T, cluster_cols=F, annotation_col=df)
dev.off()
```


<!-- Characterise Latent Factor 2 -->
```{r}
# Plot top weights of RNA view
pdf(file=paste0(io$outdir,"/topWeights_Factor2.pdf"), useDingbats = F, width=4.5, height=1.5)
plotWeights(model, view="RNA expression", factor=2, nfeatures = 0,
            manual=list(c("Krt8","Cald1","Anxa5","Tagln","Ahnak","Dsp","Anxa3","Krt19")), abs=T)
dev.off()


# Plot top features heatmap, after removing 2i cells
# df <- data.frame(
#   row.names = sample_metadata_filt$sample,
#   culture = sample_metadata_filt$culture
# )
# pdf(file=paste0(io$outdir,"/Heatmap_Factor2.pdf"), width=6, height=4, onefile = F)
# plotDataHeatmap(model, view="RNA expression", factor=2, features = 60, annotation_row=df, show_colnames=F, show_rownames=F, sortSamples=T, cluster_rows=F)
# dev.off()

# Plot top features heatmap, ignoring cells not important in Factor 2
Z <- getFactors(model, factors="2")
samples <- Z[sample_metadata_filt[culture!="2i" & pass_rnaQC==T,sample],]
samples <- names(samples)[order(samples)]
features <- getWeights(model,views = "RNA expression", factors = 2, as.data.frame = T) %>% as.data.table %>% 
  # .[feature %in% c("Krt8","Wls","Cyba","Dppa5a","Zfp981")] %>% .$feature
  .[,value:=abs(value)] %>% setkey(value) %>% tail(n=23) %>% .[!feature%in%c("Cmtm3","Gjb3","Emb")] %>% .$feature

df <- data.frame(
  row.names = samples,
  Factor2 = Z[samples,]
)

tmp <- model@TrainData$`RNA expression`[features,samples]

pdf(file=paste0(io$outdir,"/Heatmap_Factor2.pdf"), width=6, height=4, onefile = F)
pheatmap::pheatmap(tmp, show_rownames = T, show_colnames = F, cluster_cols = F, annotation_col=df)
dev.off()
```


<!-- Bivariate plot of the weights -->
```{r}
W <- getWeights(model, factors=c("1","2"), views=c("RNA expression"), as.data.frame = T) %>%
  dcast(feature+view~factor)
colnames(W) <- c("feature","view","factorX","factorY")
W$factorX <- abs(W$factorX)/max(abs(W$factorX))
W$factorY <- abs(W$factorY)/max(abs(W$factorY))
# W$factorX <- W$factorX/max(abs(W$factorX))
# W$factorY <- W$factorY/max(abs(W$factorY))


W$col <- abs(W$factorX)+abs(W$factorY)

# Removed: Anxa3, Tagln
W$interesting_features_Factor1 <- W$feature %in% c("Zfp42","Esrrb","Morc1","Fbxo15","Jam2","Klf4","Tcl1","Tbx3","Tex19.1","Rpl10l")
W$interesting_features_Factor2 <- W$feature %in% c("Krt8","Cald1","Anxa5","Ahnak","Dsp","Krt19","Utf1","Ina","Cmtm3","Cmtm7")
W$interesting_features_Factor12 <- W$feature %in% c("Tagln","Tdh","Rex2","Wnt3","Tpm4","Igfbp3")


p <- ggplot(W, aes(x=factorX, y = factorY)) + 
  geom_point(color="lightgrey", size=0.3) +
  geom_point(aes(x=factorX, y = factorY), data=W[W$interesting_features_Factor1==TRUE,], color="orange") +
  geom_point(aes(x=factorX, y = factorY), data=W[W$interesting_features_Factor2==TRUE,], color="yellowgreen") +
  geom_point(aes(x=factorX, y = factorY), data=W[W$interesting_features_Factor12==TRUE,], color="blue") +
  scale_colour_gradient(low = "lightgrey", high = "black") +
  ggrepel::geom_text_repel(data = W[W$interesting_features_Factor1==TRUE,], aes(label = feature), color="orange", size=5,
                         segment.alpha=0.3, segment.color="orange", segment.size=0.6, box.padding = unit(0.5, "lines")) +
  ggrepel::geom_text_repel(data = W[W$interesting_features_Factor2==TRUE,], aes(label = feature), color="yellowgreen", size=5,
                         segment.alpha=0.3, segment.color="yellowgreen", segment.size=0.6, box.padding = unit(0.5, "lines")) +
  ggrepel::geom_text_repel(data = W[W$interesting_features_Factor12==TRUE,], aes(label = feature), color="blue", size=5,
                         segment.alpha=0.3, segment.color="yellowgreen", segment.size=0.6, box.padding = unit(0.5, "lines")) +
  xlab("Loadings Factor 1") + ylab("Loadings Factor 2") +
  theme(
    plot.margin = margin(20, 20, 10, 10),
    axis.text = element_text(size = rel(1), color = "black"),
    axis.title = element_text(size = 16),
    axis.title.y = element_text(size = rel(1.1), margin = margin(0, 10, 0, 0)),
    axis.title.x = element_text(size = rel(1.1), margin = margin(10, 0, 0, 0)),
    axis.line = element_line(color = "black", size = 0.5),
    axis.ticks = element_line(color = "black", size = 0.5),
    panel.border = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.background = element_blank(),
    legend.position = "none",
    legend.key = element_rect(fill = "white"),
    legend.text = element_text(size = 16),
    legend.title = element_text(size =16)
  )
p

# pdf(file=paste0(io$outdir,"/weightsFactors12.pdf"), width=6, height=4, onefile = F)
# print(p)
# dev.off()
```

<!-- Characterise Latent Factor 3 -->
```{r}

# Correlation with cellular detection rate
cor ( getFactors(model, include_intercept = F), colMeans(model@TrainData$`RNA expression`,na.rm=T), use="complete.obs")
plotWeights(model, view="RNA expression", factor=3, nfeatures = 30)

foo <- data.frame(
  factor = getFactors(model, factors=3, include_intercept = F)[,1],
  cdr = colMeans(model@TrainData$`RNA expression`>0,na.rm=T)
)

p <- ggplot(foo, aes_string(x = "factor", y = "cdr")) + 
  geom_point() + xlab("Factor 3") + ylab("Cellular Detection Rate") +
  stat_smooth(method="lm") +
  theme(
    plot.margin = margin(20, 20, 10, 10), 
    axis.text = element_text(size = rel(1), color = "black"), 
    axis.title = element_text(size = 16), 
    axis.title.y = element_text(size = rel(1.1), margin = margin(0, 10, 0, 0)), 
    axis.title.x = element_text(size = rel(1.1), margin = margin(10, 0, 0, 0)), 
    axis.line = element_line(color = "black", size = 0.5), 
    axis.ticks = element_line(color = "black", size = 0.5),
    panel.border = element_blank(), 
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(), 
    panel.background = element_blank(),
    legend.key = element_rect(fill = "white"),
    legend.text = element_text(size = 16),
    legend.title = element_text(size =16)
  )
p

# pdf(file=paste0(io$outdir,"/cdrFactor3.pdf"), width=6, height=4, onefile = F)
# print(p)
# dev.off()

```

<!-- Compare results to Pseudotime solution -->
```{r}
library(destiny)
rna_matrix <- getTrainData(model, views="RNA expression")[[1]] %>% t
samples <- sample_metadata_filt[pass_rnaQC==TRUE,sample]
rna_matrix <- rna_matrix[samples,]

dm <- DiffusionMap(rna_matrix)

# plot(
#     eigenvectors(dm)[,1],
#     eigenvectors(dm)[,2],
#     xlab="Diffusion component 1",
#     ylab="Diffusion component 2",
#     col = c("red","green")[as.numeric(factor(sce$lineage))],
#     pch = c(16,17)[as.numeric(factor(sce$lineage))]
# )

tmp <- eigenvectors(dm)[,c(1,2)] %>% as.data.table %>% .[,sample:=rownames(rna_matrix)] %>% 
  merge(sample_metadata,by="sample")

p <- ggplot(tmp, aes(x=DC1, y=DC2)) +
  geom_point(aes(color=culture), size=2.5) +
  xlab("DC1") + ylab("DC2") +
  theme_bw()

pdf(file=paste0(io$outdir,"/destiny.pdf"), width=6, height=4)
print(p)
dev.off()
```


<!-- Compare results to PCA solution -->
<!-- ```{r} -->
<!-- matrix <- getTrainData(model, views="RNA expression")[[1]] %>% t -->
<!-- # matrix <- getTrainData(model, views="met_prom_2000_2000_noncgi")[[1]] %>% t -->

<!-- matrix_i <- impute(matrix, margin=2) -->
<!-- pc.out <- pcaMethods::pca(matrix_i, method="svd", center=TRUE, scale="none", nPcs=2) -->

<!-- tmp <- pc.out@scores %>% as.data.table %>% .[,sample:=rownames(matrix_i)] %>% -->
<!--   merge(sample_metadata,by="sample") -->

<!-- p <- ggplot(tmp, aes(x=PC1, y=PC2)) + -->
<!--   geom_point(aes(color=culture), size=2.5) + -->
<!--   xlab("PC1") + ylab("PC2") + -->
<!--   theme_bw() -->
<!-- print(p) -->

<!-- # pdf(file=paste0(io$outdir,"/Heatmap_Factor1.pdf"), width=6, height=4, onefile = F) -->
<!-- ``` -->



<!-- SNF solution -->
```{r}

library(SNFtool)

# set all the parameters:
K = 20;		# number of neighbors, usually (10~30)
alpha = 0.5;  	# hyperparameter, usually (0.3~0.8)
iter = 10; 	# Number of Iterations, usually (10~20)


# SNF does not cope with samples missing assays, use only overlapping samples
sample_metadata_filt <- sample_metadata[pass_rnaQC==TRUE & pass_metQC==TRUE]
## Calculate distance matrices(here we calculate Euclidean Distance, you can use other distance, e.g,correlation)

## If the data are all continuous values, we recommend the users to perform standard normalization before using SNF, though it is optional depending on the data the users want to use.  

all_matrix_list2 <- lapply(all_matrix_list, function(x) t(x[,sample_metadata_filt$sample]) )
all_matrix_list2$rna <- standardNormalization(all_matrix_list2$rna)

# Data2 = standardNormalization(Data2);


## Calculate the pair-wise distance; If the data is continuous, we recommend to use the function "dist2" as follows; if the data is discrete, we recommend the users to use ""
Dist1 = dist2(all_matrix_list2$rna,all_matrix_list2$rna)
Dist2 <- dist(all_matrix_list2$met_CGI, method="binary") %>% as.matrix
Dist2[is.na(Dist2)] <- 0.0
Dist3 <- dist(all_matrix_list2$met_activeEnhancers_ENCODE, method="euclidean") %>% as.matrix
Dist3[is.na(Dist3)] <- 0.0
Dist4 <- dist(all_matrix_list2$met_prom_2000_2000, method="euclidean") %>% as.matrix
Dist4[is.na(Dist4)] <- 0.0

# Construct similarity graphs
W1 = affinityMatrix(Dist1, K, alpha)
W2 = affinityMatrix(Dist2, K, alpha)
W3 = affinityMatrix(Dist3, K, alpha)
W4 = affinityMatrix(Dist4, K, alpha)

# These similarity graphs have complementary information about clusters.
# displayClusters(W1,sample_metadata_filt$culture);
# displayClusters(W2,sample_metadata_filt$culture);

## Fuse all the graphs and compute overall matrix by similarity network fusion:
W = SNF(list(W1,W2,W3,W4), K, iter)
# displayClusters(W, sample_metadata_filt$culture)

## Spectral clustering
group = spectralClustering(W, K=2)
# sample_metadata_filt$culture[group-1]

# Plot similarity matrix
foo <- W / rowSums(W)
foo = foo + t(foo)
diag(foo) <- 0
rownames(foo) <- colnames(foo) <- sample_metadata_filt$sample

row_df <- data.frame(
  row.names = sample_metadata_filt$sample,
  culture = sample_metadata_filt$culture
)

col_df <- data.frame(
  row.names = sample_metadata_filt$sample,
  cluster = as.factor(group)
)

annotation_colors <- list(
  "culture"=c("Serum"="orangered3","2i"="lightsalmon")
)

pdf(file=paste0(io$outdir,"/SNFHeatMap.pdf"), height=5, width=7.5)
pheatmap::pheatmap(foo, annotation_row=row_df, annotation_colors=annotation_colors, show_rownames = F, show_colnames = F)
dev.off()
```


<!-- Plot coordinated transcriptomic and epigenetic changes in Factor 1 -->

RNA
```{r}

# Extract genes with top weights
top <- getWeights(model, views="RNA expression", factors = "1")[[1]][,1]
top_positive <- names(tail(sort(abs(top[top>0])), n=15))
top_negative <- names(tail(sort(abs(top[top<0])), n=15))

# Extract factor values
Z <- getFactors(model, factors="1", as.data.frame = TRUE)
rna_top <- getTrainData(model, views="RNA expression", 
                        features=list(c(top_positive,top_negative)), as.data.frame = TRUE)
rna_top$pos <- rna_top$feature %in% top_positive

foo <- merge(Z,rna_top, by="sample")

p <- ggplot(foo, aes(x=value.x, y = value.y)) + 
  stat_smooth(aes(group=feature, color=pos), method="loess", se=F, size=0.75) +
  xlab("Factor 1") + ylab("RNA expression") +
  theme_bw() +
  theme(
    # plot.margin = margin(20, 20, 10, 10),
    axis.text = element_text(size = rel(1), color = "black"),
    axis.title = element_text(size = 16),
    axis.title.y = element_text(size = rel(1.1), margin = margin(0, 10, 0, 0)),
    axis.title.x = element_text(size = rel(1.1), margin = margin(10, 0, 0, 0)),
    axis.line = element_line(color = "black", size = 0.5),
    axis.ticks = element_line(color = "black", size = 0.5),
    # panel.border = element_blank(),
    # panel.grid.major = element_blank(),
    # panel.grid.minor = element_blank(),
    # panel.background = element_blank(),
    legend.position = "none",
    legend.key = element_rect(fill = "white"),
    legend.text = element_text(size = 16),
    legend.title = element_text(size =16)
  )
print(p)
```

```{r}

# plotDataScatter(model, view="RNA expression", factor = "1", 
#                 features = c("Zfp42","Esrrb","Morc1","Fbxo15","Jam2","Klf4","Tcl1","Tbx3","Tex19.1","Rpl10l"))
# 
# plotDataScatter(model, view="RNA expression", factor = "1", 
#                 features = c("Krt8","Cald1","Anxa5","Ahnak","Dsp","Krt19","Utf1","Ina","Cmtm3","Cmtm7"))
# 
# plotDataScatter(model, view="RNA expression", factor = "1", 
#                 features = c("Tagln","Tdh","Rex2","Wnt3","Tpm4","Igfbp3"))
# 
# W$interesting_features_Factor12 <- W$feature %in% 

```


Methylation
```{r}

# Extract CpG sites with top weights
top <- getWeights(model, views="Met Promoters", factors = "1")[[1]][,1]
top_positive <- names(tail(sort(abs(top[top>0])), n=50))
# top_negative <- names(tail(sort(abs(topgenes[topgenes<0])), n=15))

# Extract factor values
Z <- getFactors(model, factors="1", as.data.frame = T)
met_top <- getTrainData(model, views="Met Promoters", 
                        features=list(top_positive), as.data.frame = T)
met_top$pos <- met_top$feature %in% top_positive

foo <- merge(Z,met_top, by="sample") %>% .[complete.cases(.),]

p <- ggplot(foo, aes(x=value.x, y = value.y)) + 
  stat_smooth(aes(group=feature, color=pos), method="loess", se=F, size=0.5) +
  xlab("Factor 1") + ylab("DNA methylation rate") +
  theme_bw() +
  theme(
    # plot.margin = margin(20, 20, 10, 10),
    axis.text = element_text(size = rel(1), color = "black"),
    axis.title = element_text(size = 16),
    axis.title.y = element_text(size = rel(1.1), margin = margin(0, 10, 0, 0)),
    axis.title.x = element_text(size = rel(1.1), margin = margin(10, 0, 0, 0)),
    axis.line = element_line(color = "black", size = 0.5),
    axis.ticks = element_line(color = "black", size = 0.5),
    # panel.border = element_blank(),
    # panel.grid.major = element_blank(),
    # panel.grid.minor = element_blank(),
    # panel.background = element_blank(),
    legend.position = "none",
    legend.key = element_rect(fill = "white"),
    legend.text = element_text(size = 16),
    legend.title = element_text(size =16)
  )
print(p)

pdf(file=paste0(io$outdir,"/MethylationChanges.pdf"), height=5, width=7.5)
print(p)
dev.off()
```

